README.md
TCP Socket Flutter
===========

* [Introduction](#introduction)
* [Set up](#setup)
* [License and contributors](#license-and-contributors)

Introduction
------------

This is a Flutter package that allows you to connect to a socket over the net. All data is then read using UTF-8.

Setup
-------
Config
```Dart
  TCPSocketSetUp.setConfig(
    const SocketConfig(
      port: 8000,
      numberSplit: 10000,
      timeoutEachTimesSendData: Duration(milliseconds: 50),
    ),
  );
```
Device Info
```Dart
await TCPSocketSetUp.init();
```
Server
```Dart
  final TCPSocketServer _server = TCPSocketServer();
  final result = await _server.initServer(
    onData: (ip, sourcePort, event) {
        print('Server receive data from: $ip:$sourcePort');
        print('Server receive data: $event');
    },
    onDone: (ip, sourcePort) {},
    onError: (error, ip, sourcePort) {},
  );
  print(result);
```
Server send data
```Dart
  await _server.sendData(
    FormDataSending(
      type: 'Server send info',
      data: getRandomString(1000000),
    ),
  );
```
Client
```Dart
  final TCPSocketClient _client = TCPSocketClient();
  final result = await _client.connectToServer(
    '192.168.0.101',
    onData: (event) {
      print('Client receive data: $event');
    },
    onDone: () {},
    onError: (error) {},
  );
  print('Connect to server $result');
```
Client send data
```Dart
  await _client.sendData(
    FormDataSending(
      type: 'Client send info',
      data: getRandomString(1000000),
    ),
  );
```

License and contributors
------------------------

* The MIT License, see [LICENSE](https://github.com/nghetien/tcp_socket_flutter/blob/main/LICENSE).
* For contributors, see [AUTHORS](https://github.com/nghetien/tcp_socket_flutter/blob/main/AUTHORS).

CHANGELOG.md
## 0.0.5

* Send utf8 character

## 0.0.4

* Fix bug don't remove SocketConnection when id of new connecting is same as id into map of server(await checkExistAndRemoveSocketConnection)

## 0.0.3

* Move streamDeviceInfo, listenerListSocketConnection to broadcast().

## 0.0.2

* Fix discoverServerIP, add more info to DeviceInfo and ServerInfo, fix send data success.

## 0.0.1

* TCP socket: start server, connect server, send big data.


example/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:tcp_socket_flutter/tcp_socket_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TCPSocketServer _server = TCPSocketServer();
  final TCPSocketClient _client = TCPSocketClient();

  final _chars =
      'AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz1234567890';
  final Random _rnd = Random();

  String getRandomString(int length) => String.fromCharCodes(Iterable.generate(
      length, (_) => _chars.codeUnitAt(_rnd.nextInt(_chars.length))));

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            MaterialButton(
              color: Colors.blue,
              onPressed: () async {
                TCPSocketSetUp.setConfig(
                  const SocketConfig(
                    port: 8000,
                    numberSplit: 10000,
                    timeoutEachTimesSendData: Duration(milliseconds: 50),
                  ),
                );
                print(TCPSocketSetUp.config.toJson());
              },
              child: const Text('Set port 8000'),
            ),
            const SizedBox(
              height: 20,
            ),
            MaterialButton(
              color: Colors.blue,
              onPressed: () async {
                await TCPSocketSetUp.init();
                print('Get device info done');
              },
              child: const Text('Get info IP'),
            ),
            const SizedBox(
              height: 20,
            ),
            const Text('Server ------------'),
            const SizedBox(
              height: 20,
            ),
            MaterialButton(
              color: Colors.blue,
              onPressed: () async {
                final result = await _server.initServer(
                  onData: (ip, sourcePort, event) {
                    print('Server receive data from: $ip:$sourcePort');
                    print('Server receive data: $event');
                  },
                  onDone: (ip, sourcePort) {},
                  onError: (error, ip, sourcePort) {},
                );
                print(result);
                print('Run server thành công');
              },
              child: const Text('Start Server'),
            ),
            MaterialButton(
              color: Colors.blue,
              onPressed: () async {
                await _server.sendData(
                  FormDataSending(
                    type: 'Server send info',
                    data: getRandomString(1000000),
                  ),
                );
              },
              child: const Text('Server send data'),
            ),
            const SizedBox(
              height: 20,
            ),
            const Text('Client ------------'),
            const SizedBox(
              height: 20,
            ),
            MaterialButton(
              color: Colors.blue,
              onPressed: () async {
                final result = await _client.connectToServer(
                  '192.168.0.101',
                  onData: (event) {
                    print('Client receive data: $event');
                  },
                  onDone: () {},
                  onError: (error) {},
                );
                print('Connect to server $result');
              },
              child: const Text('Connect to server'),
            ),
            const SizedBox(
              height: 20,
            ),
            MaterialButton(
              color: Colors.blue,
              onPressed: () async {
                await _client.sendData(
                  FormDataSending(
                    type: 'Client send info',
                    data: getRandomString(1000000),
                  ),
                );
              },
              child: const Text('Client send data'),
            ),
          ],
        ),
      ),
    );
  }
}
